workshops/face-expression-detection/README.md

```diff
@@ -0,0 +1,268 @@
+
+```python
+import cv2
+
+with open('anger.txt','r') as f:
+    images = [line.strip() for line in f]
+
+face_detector = cv2.CascadeClassifier('haarcascade_frontalface_default.xml')
+
+# For each Emotion, enter one numeric face id
+face_id = input('\n Enter Emotion id end press <return> ==>  ')
+
+count = 0
+
+for image in images:
+    img = cv2.imread("data_set/anger/"+image)
+    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+    faces = face_detector.detectMultiScale(gray, 1.3, 5)
+
+    for (x,y,w,h) in faces:
+
+        cv2.rectangle(img, (x,y), (x+w,y+h), (255,0,0), 2)     
+        count += 1
+
+        # Save the captured image into the datasets folder
+        cv2.imwrite("dataset/User." + str(face_id) + '.' + str(count) + ".jpg", gray[y:y+h,x:x+w])
+
+print("\n Done creating face data")
+```
+
+Here, there will be an input  option when you run the code. Now, we will be starting with **0 (Zero)** for anger, **1** for happy, **2** for sad, and so on. This is because when the image naming starts, all the images will be assigned a definite name instead of something random. This step is just to ease in the process of training. The above code is for images with **anger** as their expression. You can do the same for other expressions as well.
+
+## Training
+
+Let's start with training the machine. For this, we'll first create a folder named **trainer**.
+
+```
+%mkdir trainer
+```
+
+Next, we create another code snippet and `import cv2, Numpy, PIL, and os` for training purposes. We then set the path for the face image database to be **dataset**. Then we create a function to get the images and label the data, lastly saving the trained model into the directory **trainer** and named as `trainer.yml`. We will also be printing the number of emotions trained so that we can keep a track of what emotions were actually trained and if we missed any.
+
+```python
+import cv2
+import numpy as np
+from PIL import Image
+import os
+
+# Path for face image database
+path = 'dataset'
+
+recognizer = cv2.face.LBPHFaceRecognizer_create()
+detector = cv2.CascadeClassifier("haarcascade_frontalface_default.xml");
+
+# function to get the images and label data
+def getImagesAndLabels(path):
+
+    imagePaths = [os.path.join(path,f) for f in os.listdir(path)]     
+    faceSamples=[]
+    ids = []
+
+    for imagePath in imagePaths:
+
+        PIL_img = Image.open(imagePath).convert('L') # convert it to grayscale
+        img_numpy = np.array(PIL_img,'uint8')
+
+        id = int(os.path.split(imagePath)[-1].split(".")[1])
+        faces = detector.detectMultiScale(img_numpy)
+
+        for (x,y,w,h) in faces:
+            faceSamples.append(img_numpy[y:y+h,x:x+w])
+            ids.append(id)
+
+    return faceSamples,ids
+
+print ("\n [INFO] Training faces....")
+faces,ids = getImagesAndLabels(path)
+recognizer.train(faces, np.array(ids))
+
+# Save the model into trainer/trainer.yml
+recognizer.write('trainer/trainer.yml') 
+
```

Comment:
not a good practice to keep a bigg chunk of code for the workshop-doers to write, try splitting them to make them small and also add an elaborate explanation